source/js/nav-newsletter.js

```diff
@@ -0,0 +1,162 @@
+import React from "react";
+import ReactGA from "react-ga";
+import ReactDOM from "react-dom";
+import JoinUs from "./components/join/join.jsx";
+
+let primaryNav,
+  narrowMenuContainer,
+  wideMenuContainer,
+  container,
+  buttonDesktop,
+  buttonMobile,
+  joinUs,
+  buttonDismiss;
+
+class NavNewsletter {
+  constructor() {
+    this.isShown = false;
+    this.form = null;
+  }
+
+  getShownState() {
+    return this.isShown;
+  }
+
+  // Reset form
+  resetForm() {
+    this.form.reset();
+
+    let handleTransitionend = () => {
+      container.removeEventListener("transitionend", handleTransitionend);
+      buttonDismiss.textContent = "No thanks";
+    };
+    container.addEventListener("transitionend", handleTransitionend);
+  }
+
+  // For desktop+ version:
+  // transition section to its close state,
+  // remove the global 'closeFormClickHandler' click event handler
+  // and reset the form
+  closeDesktopNewsletter(event) {
+    container.classList.remove("expanded");
+    buttonDesktop.classList.remove("active");
+    document.removeEventListener("click", this.closeFormClickHandler);
+    document.removeEventListener("scroll", this.closeFormClickHandler);
+    this.resetForm();
+    this.isShown = false;
+  }
+
+  // For desktop+ version:
+  // transition newsletter section to its expanded state
+  expandDesktopNewsletter(event) {
+    container.style.top = `${primaryNav.offsetHeight}px`;
+    container.classList.add("expanded");
+    buttonDesktop.classList.add("active");
+    document.addEventListener(`click`, this.closeFormClickHandler);
+    document.addEventListener("scroll", this.closeFormClickHandler);
+    this.isShown = true;
+  }
+
+  // For mobile version:
+  // transition newsletter section to its close state,
+  // remove the global 'closeFormClickHandler' click event handler,
+  // and reset the form
+  closeMobileNewsletter() {
+    narrowMenuContainer.classList.remove("d-none");
+    container.classList.remove("faded-in");
+    this.resetForm();
+    this.isShown = false;
+  }
+
+  // For mobile version:
+  // transition section to its expanded state
+  expandMobileNewsletter() {
+    narrowMenuContainer.classList.add(`d-none`);
+    container.classList.add("faded-in");
+    this.isShown = true;
+  }
+
+  // For desktop+ version:
+  // create click handler to detect clicking event outside of the newsletter section
+  closeFormClickHandler(event) {
+    // close newsletter section if clicking anywhere outside of the section
+    if (!container.contains(event.target) && event.target !== container) {
+      navNewsletter.closeDesktopNewsletter();
+    }
+  }
+
+  checkDomNodes() {
+    let allNodesExisted = false;
+
+    try {
```

Comment:
Note that none of the code in this try block will ever throw, so try/catch wrapping will not do anything meaningful. We can instead rewrite this function to something quite a bit smaller by relying on [Array.every](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every):
```javascript
/**
 * Find and bind all necessary DOM nodes, returning "false" if not all DOM nodes were found.
 */
checkDomNodes() {
  return Object.keys(elements).every(key => {
    // if this element already resolved to a DOM node, move on to the next
    let value = elements[key];
    if (typeof value !== "string") return true;

    // find this DOM node, and report on the result (binding it if found for later use)
    let element = document.querySelector(value);
    if (element) elements[key] = element;
    return !!element;
  });
}
```

Reply:
> Note that none of the code in this try block will ever throw, so try/catch wrapping will not do anything meaningful.

Ah I thought `buttonDesktop = wideMenuContainer.querySelector(".btn-newsletter");` would throw error when `wideMenuContainer` is `null`?

Anyways I will update my PR based on the code snippet above.

Reply:
ahhh missed that - we can turn that into a "full" query selector though and then with the qs mapping object that won't be a problem.